Source/SwiftLintFramework/Rules/GenericTypeNameRule.swift

```diff
@@ -0,0 +1,235 @@
+                     "uppercase character and span between 1 and 20 characters in length.",
+        nonTriggeringExamples: [
+            "func foo<T>() {}\n",
+            "func foo<T>() -> T {}\n",
+            "func foo<T, U>(param: U) -> T {}\n",
+            "func foo<T: Hashable, U: Rule>(param: U) -> T {}\n",
+            "struct Foo<T> {}\n",
+            "class Foo<T> {}\n",
+            "enum Foo<T> {}\n",
+            "func run(_ options: NoOptions<CommandantError<()>>) {}\n",
+            "func foo(_ options: Set<type>) {}\n",
+            "func < <T: Comparable>(lhs: T?, rhs: T?) -> Bool\n",
+            "func configureWith(data: Either<MessageThread, (project: Project, backing: Backing)>)\n",
+            "typealias StringDictionary<T> = Dictionary<String, T>\n",
+            "typealias BackwardTriple<T1, T2, T3> = (T3, T2, T1)\n",
+            "typealias DictionaryOfStrings<T : Hashable> = Dictionary<T, String>\n"
+        ],
+        triggeringExamples: [
+            "func foo<↓T_Foo>() {}\n",
+            "func foo<T, ↓U_Foo>(param: U_Foo) -> T {}\n",
+            "func foo<↓\(String(repeating: "T", count: 21))>() {}\n",
+            "func foo<↓type>() {}\n",
+            "typealias StringDictionary<↓T_Foo> = Dictionary<String, T_Foo>\n",
+            "typealias BackwardTriple<T1, ↓T2_Bar, T3> = (T3, T2_Bar, T1)\n",
+            "typealias DictionaryOfStrings<↓T_Foo: Hashable> = Dictionary<T, String>\n"
+        ] + ["class", "struct", "enum"].flatMap { type in
+            [
+                "\(type) Foo<↓T_Foo> {}\n",
+                "\(type) Foo<T, ↓U_Foo> {}\n",
+                "\(type) Foo<↓T_Foo, ↓U_Foo> {}\n",
+                "\(type) Foo<↓\(String(repeating: "T", count: 21))> {}\n",
+                "\(type) Foo<↓type> {}\n"
+            ]
+        }
+    )
+
+    private let genericTypePattern = "<(\\s*\\w.*?)>"
+    private var genericTypeRegex: NSRegularExpression {
+        return regex(genericTypePattern)
+    }
+
+    public func validateFile(_ file: File) -> [StyleViolation] {
+        return validateFile(file, dictionary: file.structure.dictionary) +
+            validateGenericTypeAliases(file)
+    }
+
+    public func validateFile(_ file: File,
+                             kind: SwiftDeclarationKind,
+                             dictionary: [String: SourceKitRepresentable]) -> [StyleViolation] {
+        let types = genericTypesForType(file, kind: kind, dictionary: dictionary) +
+                    genericTypesForFunction(file, kind: kind, dictionary: dictionary)
+
+        return types.flatMap { validateName(name: $0.0, file: file, offset: $0.1) }
+    }
+
+    private func validateGenericTypeAliases(_ file: File) -> [StyleViolation] {
+        let pattern = "typealias\\s+.+?" + genericTypePattern + "\\s*="
+        return file.matchPattern(pattern).flatMap { (range, tokens) -> [(String, Int)] in
+            guard tokens.count > 1, tokens.first == .keyword,
```

Comment:
I think the `tokens.count > 1` check is superfluous and covered by the subsequent two checks.